src/PrestaShopBundle/Routing/Converter/CacheProvider.php

```diff
@@ -0,0 +1,129 @@
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\Routing\Converter;
+
+use Symfony\Component\Cache\Adapter\AdapterInterface;
+
+/**
+ * Class CacheProvider.
+ */
+class CacheProvider extends AbstractLegacyRouteProvider
+{
+    /**
+     * @var AdapterInterface
+     */
+    private $cache;
+
+    /**
+     * @var LegacyRouteProviderInterface
+     */
+    private $legacyRouteProvider;
+
+    /**
+     * @var CacheKeyGeneratorInterface
+     */
+    private $cacheKeyGenerator;
+
+    /**
+     * @var LegacyRoute[]
+     */
+    private $legacyRoutes;
+
+    public function __construct(
+        LegacyRouteProviderInterface $legacyRouteProvider,
+        AdapterInterface $cache,
+        CacheKeyGeneratorInterface $cacheKeyGenerator
+    ) {
+        $this->legacyRouteProvider = $legacyRouteProvider;
+        $this->cache = $cache;
+        $this->cacheKeyGenerator = $cacheKeyGenerator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getLegacyRoutes()
+    {
+        if (null === $this->legacyRoutes) {
+            $cacheItem = $this->cache->getItem($this->cacheKeyGenerator->getCacheKey());
+            if (!$cacheItem->isHit()) {
+                $this->legacyRoutes = $this->legacyRouteProvider->getLegacyRoutes();
+                $cacheItem->set($this->serializeLegacyRoutes($this->legacyRoutes));
+                $this->cache->save($cacheItem);
+            } else {
+                $this->legacyRoutes = $this->unserializeLegacyRoutes($cacheItem->get());
+            }
+        }
+
+        return $this->legacyRoutes;
+    }
+
+    /**
+     * @param LegacyRoute[] $legacyRoutes
+     *
+     * @return string
+     */
+    private function serializeLegacyRoutes(array $legacyRoutes)
+    {
+        $flattenRoutes = [];
+        /** @var LegacyRoute $legacyRoute */
+        foreach ($legacyRoutes as $legacyRoute) {
+            $legacyLinks = [];
+            foreach ($legacyRoute->getLegacyLinks() as $legacyLink) {
+                $legacyLinks[] = $legacyLink['controller'] . (!empty($legacyLink['action']) ? ':' . $legacyLink['action'] : '');
```

Comment:
Did you consider using `sprintf()` to increase readability and reduce errors risk ?